// sc-errors/index.js
var cycle = require('cycle');

var isStrict = (function () { return !this; })();

function AuthTokenExpiredError(message, expiry) {
  if (Error.captureStackTrace && !isStrict) {
    Error.captureStackTrace(this, arguments.callee);
  } else {
    this.stack = (new Error()).stack;
  }
  this.name = 'AuthTokenExpiredError';
  this.message = message;
  this.expiry = expiry;
};
AuthTokenExpiredError.prototype = Object.create(Error.prototype);


function AuthTokenInvalidError(message) {
  if (Error.captureStackTrace && !isStrict) {
    Error.captureStackTrace(this, arguments.callee);
  } else {
    this.stack = (new Error()).stack;
  }
  this.name = 'AuthTokenInvalidError';
  this.message = message;
};
AuthTokenInvalidError.prototype = Object.create(Error.prototype);


function SilentMiddlewareBlockedError(message, type) {
  if (Error.captureStackTrace && !isStrict) {
    Error.captureStackTrace(this, arguments.callee);
  } else {
    this.stack = (new Error()).stack;
  }
  this.name = 'SilentMiddlewareBlockedError';
  this.message = message;
  this.type = type;
};
SilentMiddlewareBlockedError.prototype = Object.create(Error.prototype);


function InvalidActionError(message) {
  if (Error.captureStackTrace && !isStrict) {
    Error.captureStackTrace(this, arguments.callee);
  } else {
    this.stack = (new Error()).stack;
  }
  this.name = 'InvalidActionError';
  this.message = message;
};
InvalidActionError.prototype = Object.create(Error.prototype);

function InvalidArgumentsError(message) {
  if (Error.captureStackTrace && !isStrict) {
    Error.captureStackTrace(this, arguments.callee);
  } else {
    this.stack = (new Error()).stack;
  }
  this.name = 'InvalidArgumentsError';
  this.message = message;
};
InvalidArgumentsError.prototype = Object.create(Error.prototype);

function InvalidOptionsError(message) {
  if (Error.captureStackTrace && !isStrict) {
    Error.captureStackTrace(this, arguments.callee);
  } else {
    this.stack = (new Error()).stack;
  }
  this.name = 'InvalidOptionsError';
  this.message = message;
};
InvalidOptionsError.prototype = Object.create(Error.prototype);


function InvalidMessageError(message) {
  if (Error.captureStackTrace && !isStrict) {
    Error.captureStackTrace(this, arguments.callee);
  } else {
    this.stack = (new Error()).stack;
  }
  this.name = 'InvalidMessageError';
  this.message = message;
};
InvalidMessageError.prototype = Object.create(Error.prototype);


function SocketProtocolError(message, code) {
  if (Error.captureStackTrace && !isStrict) {
    Error.captureStackTrace(this, arguments.callee);
  } else {
    this.stack = (new Error()).stack;
  }
  this.name = 'SocketProtocolError';
  this.message = message;
  this.code = code;
};
SocketProtocolError.prototype = Object.create(Error.prototype);


function ServerProtocolError(message) {
  if (Error.captureStackTrace && !isStrict) {
    Error.captureStackTrace(this, arguments.callee);
  } else {
    this.stack = (new Error()).stack;
  }
  this.name = 'ServerProtocolError';
  this.message = message;
};
ServerProtocolError.prototype = Object.create(Error.prototype);

function HTTPServerError(message) {
  if (Error.captureStackTrace && !isStrict) {
    Error.captureStackTrace(this, arguments.callee);
  } else {
    this.stack = (new Error()).stack;
  }
  this.name = 'HTTPServerError';
  this.message = message;
};
HTTPServerError.prototype = Object.create(Error.prototype);


function ResourceLimitError(message) {
  if (Error.captureStackTrace && !isStrict) {
    Error.captureStackTrace(this, arguments.callee);
  } else {
    this.stack = (new Error()).stack;
  }
  this.name = 'ResourceLimitError';
  this.message = message;
};
ResourceLimitError.prototype = Object.create(Error.prototype);


function TimeoutError(message) {
  if (Error.captureStackTrace && !isStrict) {
    Error.captureStackTrace(this, arguments.callee);
  } else {
    this.stack = (new Error()).stack;
  }
  this.name = 'TimeoutError';
  this.message = message;
};
TimeoutError.prototype = Object.create(Error.prototype);


function BrokerError(message) {
  if (Error.captureStackTrace && !isStrict) {
    Error.captureStackTrace(this, arguments.callee);
  } else {
    this.stack = (new Error()).stack;
  }
  this.name = 'BrokerError';
  this.message = message;
};
BrokerError.prototype = Object.create(Error.prototype);


function ProcessExitError(message, code) {
  if (Error.captureStackTrace && !isStrict) {
    Error.captureStackTrace(this, arguments.callee);
  } else {
    this.stack = (new Error()).stack;
  }
  this.name = 'ProcessExitError';
  this.message = message;
  this.code = code;
};
ProcessExitError.prototype = Object.create(Error.prototype);


function UnknownError(message) {
  if (Error.captureStackTrace && !isStrict) {
    Error.captureStackTrace(this, arguments.callee);
  } else {
    this.stack = (new Error()).stack;
  }
  this.name = 'UnknownError';
  this.message = message;
};
UnknownError.prototype = Object.create(Error.prototype);


// Expose all error types

module.exports = {
  AuthTokenExpiredError: AuthTokenExpiredError,
  AuthTokenInvalidError: AuthTokenInvalidError,
  SilentMiddlewareBlockedError: SilentMiddlewareBlockedError,
  InvalidActionError: InvalidActionError,
  InvalidArgumentsError: InvalidArgumentsError,
  InvalidOptionsError: InvalidOptionsError,
  InvalidMessageError: InvalidMessageError,
  SocketProtocolError: SocketProtocolError,
  ServerProtocolError: ServerProtocolError,
  HTTPServerError: HTTPServerError,
  ResourceLimitError: ResourceLimitError,
  TimeoutError: TimeoutError,
  BrokerError: BrokerError,
  ProcessExitError: ProcessExitError,
  UnknownError: UnknownError
};

module.exports.socketProtocolErrorStatuses = {
  1001: 'Socket was disconnected',
  1002: 'A WebSocket protocol error was encountered',
  1003: 'Server terminated socket because it received invalid data',
  1005: 'Socket closed without status code',
  1006: 'Socket hung up',
  1007: 'Message format was incorrect',
  1008: 'Encountered a policy violation',
  1009: 'Message was too big to process',
  1010: 'Client ended the connection because the server did not comply with extension requirements',
  1011: 'Server encountered an unexpected fatal condition',
  4000: 'Server ping timed out',
  4001: 'Client pong timed out',
  4002: 'Server failed to sign auth token',
  4003: 'Failed to complete handshake',
  4004: 'Client failed to save auth token',
  4005: 'Did not receive #handshake from client before timeout',
  4006: 'Failed to bind socket to message broker',
  4007: 'Client connection establishment timed out'
};

module.exports.socketProtocolIgnoreStatuses = {
  1000: 'Socket closed normally',
  1001: 'Socket hung up'
};

// Properties related to error domains cannot be serialized.
var unserializableErrorProperties = {
  domain: 1,
  domainEmitter: 1,
  domainThrown: 1
};

module.exports.dehydrateError = function (error, includeStackTrace) {
  var dehydratedError;
  if (typeof error == 'string') {
      dehydratedError = error;
  } else {
    dehydratedError = {
      message: error.message
    };
    if (includeStackTrace) {
      dehydratedError.stack = error.stack;
    }
    for (var i in error) {
      if (!unserializableErrorProperties[i]) {
        dehydratedError[i] = error[i];
      }
    }
  }
  return cycle.decycle(dehydratedError);
};

module.exports.hydrateError = function (error) {
  var hydratedError = null;
  if (error != null) {
    if (typeof error == 'string') {
      hydratedError = error;
    } else {
      hydratedError = new Error(error.message);
      for (var i in error) {
        if (error.hasOwnProperty(i)) {
          hydratedError[i] = error[i];
        }
      }
    }
  }
  return hydratedError;
};
